### DISCUSSION TITLE
Custom previews

### DISCUSSION THREAD

**Discussion by dindinet** [sveltia/sveltia-cms#153] [General]
Hey Kohei

Do you have a timeline for Custom Previews?

Will it be a lot of work?

Where in the repo is the code that does the default preview?

regards @dindinet

**Comment by kyoshino** [comment]
Hey 👋🏼 Custom Previews is https://decapcms.org/docs/customization/ right?

- `registerPreviewStyle` (CSS registration) is probably not a big deal; I’ll work on it sometime this month.
- `registerPreviewTemplate` might take a bit of time because it’s a React thing, but I’ll do it anyway before the 1.0 release planned for this Q3.

The default CSS is embedded in [`field-preview.svelte`](https://github.com/sveltia/sveltia-cms/blob/main/src/lib/components/contents/details/preview/field-preview.svelte).

**Comment by dindinet** [comment]
Hi Kohei

Specifically
**registerPreviewTemplate**
Registers a template for a folder collection or an individual file in a file collection.

**CMS.registerPreviewTemplate(name, react_component);**

You won't implement it in React though, will you?

More likely 
**CMS.registerPreviewTemplate(name, svelte_component/template);**

@dindinet

**Reply by kyoshino** [reply]
Got it! For backward compatibility, I’ll add support for preview templates, [custom widgets, and custom editor components](https://decapcms.org/docs/custom-widgets/) written in React. [Using React in Svelte](https://pandemicode.dev/using-react-within-your-svelte-applications-3b1f2a75aefc) should be possible 🙂 Maybe in the future I could add support for Svelte!

**Reply by kyoshino** [reply]
If possible, I would use [Preact](https://preactjs.com/) instead of React because JSX support is not required.

**Reply by dindinet** [reply]
I would vote for Svelte first. Surely for end users converting react templates to svelte would be less onerous than the headache of writing templates in react. Could it be something as simple as directory per page type.
For svelte. Have a directory called sveltepreviewtemplates with subfolders post,page, othertype... with an index.svelte.

something akin to the way sveltekit implements routes. ( I will read the Using React in Svelte article now :) )

**Reply by kyoshino** [reply]
I guess Svelte support is rather difficult, because Svelte components basically need to be pre-compiled. It’s technically possible to compile components on the fly using [svelte/compiler](https://svelte.dev/docs/svelte-compiler), but adding it to Sveltia makes the bundle size much larger, IIRC.

Since the main target of Sveltia CMS is existing Netlify/Decap CMS users, making their migration easier is a priority 🙂

**Reply by dindinet** [reply]
Aaagh. Svelte first. Worry about React Backwards Confusability later. :)

**Reply by dindinet** [reply]
When I have used Custom Preview Templates with Netlify CMS I have bundled them using esbuild. This is not exactly plug and play any updates to the core CMS needs a rebuild and deploy. I could never get the functionality I wanted with their example using an inline script block template.

**Reply by kyoshino** [reply]
You know I’m a [Svelte backer](https://opencollective.com/kyoshino) and I hate React, but an easy migration is a must for Sveltia. As the project [README](https://github.com/sveltia/sveltia-cms/blob/main/README.md) says, it’s designed to be a drop-in replacement for Netlify/Decap CMS. Asking existing users to rewrite their components in Svelte is not an option, unfortunately 🤣

**Reply by dindinet** [reply]
<script src="https://unpkg.com/decap-cms@^3.0.0/dist/decap-cms.js"></script>
    <script>
      var PostPreview = createClass({
        render: function() {
          var entry = this.props.entry;
          var image = entry.getIn(['data', 'image']);
           var bg = this.props.getAsset(image);
          return h('div', {},
            h('h1', {}, entry.getIn(['data', 'title'])),
            h('img', {src: bg.toString()}),
            h('div', {"className": "text"}, this.props.widgetFor('body'))
          );
        }
      });

      CMS.registerPreviewTemplate("posts", PostPreview);
    </script>

**Reply by dindinet** [reply]
Point taken. So for the time being we need to get our heads round Preact and the ideas mentioned in the Using React in Svelte article?

**Reply by kyoshino** [reply]
Actually I want to add support for [Preact](https://preactjs.com/)+[HTM](https://github.com/developit/htm) as an alternative to the `h` hell. No transpiler necessary. I used the combo for my previous project, and it worked well. It’s easy to learn for developers familiar with React. Svelte 5+ adoption will take some time.

**Reply by kyoshino** [reply]
Anyway, I need to add compatibility for the `h` thing first.

**Reply by tobiasBora** [reply]
Why not also add support for custom components? (possibly in a new command not to break compatibility) This way it would be framework-agnostic.

**Comment by jerefrer** [comment]
Hey there,

First I'd like to thank so very much everyone involved in this project 🙏
I've been following it for some time and finally got to use it for a client and so far things have been near perfect.

The only thing that I feel is missing is that there isn't a "live preview" feature where the client can see how the changes would display in the actual website. Is that what you are actually discussing here? Or is there such a feature and I missed it?

I'm a bit baffled by the number of CMSes around, and I couldn't find a single one that checks all the boxes of:
- proper i18n support
- live preview
- git-based
- framework agnostic

I have a great hopes that you could be the one 😄

Thanks for reading this, have a great day!

**Reply by kyoshino** [reply]
Sveltia CMS doesn’t have plan to support live preview out of the box, because it’s framework-specific.

**Reply by kyoshino** [reply]
Depending on your framework, if you run the CMS locally and register a preview template that embeds an `<iframe>`, the framework could give you real live preview, as per #169. 

If it’s not the case, you have to provide full HTML/CSS with a template to render a preview that mimics the actual site.

**Comment by jerefrer** [comment]
Could custom preview kind of serve this purpose by mimicking the style of
the whole page or maybe rather the style of each element?

Otherwise would there be a way to expose an api that would allow the
developer to implement live preview for his framework of choice?

Jeremy


Le dim. 12 oct. 2025 à 19:13, Kohei Yoshino ***@***.***> a
écrit :

**Reply by kyoshino** [reply]
[`CMS.registerPreviewTemplate()`](https://decapcms.org/docs/customization/#registerpreviewtemplate) serves as an API once implemented in Sveltia CMS.

**Reply by kyoshino** [reply]
Sveltia CMS has already implemented [`CMS.registerPreviewStyle()`](https://decapcms.org/docs/customization/#registerpreviewstyle), btw. In some simple cases, like a blog post, that’s enough to mimic the actual site 😉 

`CMS.registerPreviewTemplate()` will be implemented by the end of the year.

**Comment by encas-parka** [comment]
Hi !
It would be cool to be able to choose to display the widget's "summary" rather than all the data in the preview.
Thank's for your work !

**Reply by kyoshino** [reply]
You mean, you want to use the `summary` option for List/Object fields for previews? Noted for future enhancements 🙂

**Reply by encas-parka** [reply]
Yes, This would be useful to avoid overloading the preview when there are many lists/objects :)
